diff -ruN a/src/evdev.c b/src/evdev.c
--- a/src/evdev.c	2025-11-25 04:40:43.000000000 +0100
+++ b/src/evdev.c	2025-11-28 08:45:44.941437857 +0100
@@ -2016,6 +2016,12 @@
 		device->scroll.invert_horizontal_scrolling = true;
 	}
 
+	/* Enable swapping of scrolling axes when quirk is present */
+	if (evdev_device_has_model_quirk(device,
+				     QUIRK_MODEL_SWAP_SCROLLING_AXES)) {
+		device->scroll.swap_scrolling_axes = true;
+	}
+
 	return fallback_dispatch_create(&device->base);
 }
 
diff -ruN a/src/evdev-fallback.c b/src/evdev-fallback.c
--- a/src/evdev-fallback.c	2025-11-25 04:40:43.000000000 +0100
+++ b/src/evdev-fallback.c	2025-11-30 16:20:51.527323248 +0100
@@ -207,6 +207,16 @@
 		      struct evdev_device *device,
 		      uint64_t time)
 {
+	if (device->scroll.swap_scrolling_axes) {
+		int tmp = dispatch->wheel.lo_res.x;
+		dispatch->wheel.lo_res.x = dispatch->wheel.lo_res.y;
+		dispatch->wheel.lo_res.y = tmp;
+
+		tmp = dispatch->wheel.hi_res.x;
+		dispatch->wheel.hi_res.x = dispatch->wheel.hi_res.y;
+		dispatch->wheel.hi_res.y = tmp;
+	}
+
 	if (!libinput_device_has_capability(&device->base, LIBINPUT_DEVICE_CAP_POINTER))
 		return;
 
@@ -344,6 +354,7 @@
 {
 	struct libinput_device *base = &device->base;
 	struct libinput_seat *seat = base->seat;
+
 	struct device_coords point;
 	struct mt_slot *slot;
 	int seat_slot;
diff -ruN a/src/evdev.h b/src/evdev.h
--- a/src/evdev.h	2025-11-25 04:40:43.000000000 +0100
+++ b/src/evdev.h	2025-11-28 08:45:36.845751432 +0100
@@ -241,6 +241,10 @@
 		 * horizontal scrolling */
 		bool invert_horizontal_scrolling;
 
+		/* set during device init to swap the axes of
+		 * scrolling (vertical <-> horizontal) */
+		bool swap_scrolling_axes;
+
 		/* angle per REL_WHEEL click in degrees */
 		struct wheel_angle wheel_click_angle;
 
diff -ruN a/src/quirks.c b/src/quirks.c
--- a/src/quirks.c	2025-11-25 04:40:43.000000000 +0100
+++ b/src/quirks.c	2025-11-28 08:51:44.401001305 +0100
@@ -266,6 +266,8 @@
 		return "ModelHPZBookStudioG3";
 	case QUIRK_MODEL_INVERT_HORIZONTAL_SCROLLING:
 		return "ModelInvertHorizontalScrolling";
+	case QUIRK_MODEL_SWAP_SCROLLING_AXES:
+		return "ModelSwapScrollingAxes";
 	case QUIRK_MODEL_LENOVO_SCROLLPOINT:
 		return "ModelLenovoScrollPoint";
 	case QUIRK_MODEL_LENOVO_T450_TOUCHPAD:
diff -ruN a/src/quirks.h b/src/quirks.h
--- a/src/quirks.h	2025-11-25 04:40:43.000000000 +0100
+++ b/src/quirks.h	2025-11-28 08:45:55.721470644 +0100
@@ -79,6 +79,7 @@
 	QUIRK_MODEL_HP_PAVILION_DM4_TOUCHPAD,
 	QUIRK_MODEL_HP_ZBOOK_STUDIO_G3,
 	QUIRK_MODEL_INVERT_HORIZONTAL_SCROLLING,
+	QUIRK_MODEL_SWAP_SCROLLING_AXES,
 	QUIRK_MODEL_LENOVO_SCROLLPOINT,
 	QUIRK_MODEL_LENOVO_T450_TOUCHPAD,
 	QUIRK_MODEL_LENOVO_X1GEN6_TOUCHPAD,
